orttraining/orttraining/python/training/ortmodule.py

```diff
@@ -317,20 +322,24 @@ def _convert_training_graph_input_to_list(self, *inputs, **kwargs):
 
         TODO: How IO binding model inputs and outputs affects initializer copies?
 
-        ONNX Runtime forward requires an order list of:
+        ONNX Runtime forward requires an ordered list of:
             * User input: computed from forward InferenceSession
             * Initializers: computed from original PyTorch model parameters
         '''
         # User inputs
+        non_none_inputs = [inp for inp in inputs if inp is not None]
         result = []
-        for input_idx, name in enumerate(self._original_module_input_names):
+        for input_idx, name in enumerate(self._onnx_graphs_info.user_input_names):
             inp = None
-            if input_idx < len(inputs) and inputs[input_idx] is not None:
-                inp = inputs[input_idx]
+            if input_idx < len(non_none_inputs):
+                inp = non_none_inputs[input_idx]
             elif name in kwargs and kwargs[name] is not None:
                 inp = kwargs[name]
-            if inp is not None and name in self._onnx_graphs_info.user_input_names:
+            if inp is not None:
                 result.append(inp)
+            else:
+                # TODO: Re-export ONNX if any input from _onnx_graphs_info.user_input_names is None.
+                raise RuntimeError(f'Input is present in ONNX graph but not provided: {name}.')
```

Comment:
shall we also raise an exception is ```len(non_none_inputs) > len(self._onnx_graphs_info.user_input_names```?

Reply:
maybe ```len(non_none_inputs) > len(self._onnx_graphs_info.user_input_names``` is not a correct statement. What I meant was what if users provided more inputs than what was already known to the onnx graph (positional inputs+kwargs)? In that case, we would need to re-export again, right?

Reply:
I'm doing this in a separate PR that will do the re-exporting.

Reply:
Is there a ADO item in our board for this extra task?

Reply:
Yes, the item is "Re-export (invalidate) the model when the input changes".